jax/api.py

```diff
@@ -406,25 +597,57 @@ def computation_maker(*args, **kwargs):
     else:
       out_parts_flat = tuple(flatten_axes(
           "xla_computation out_parts", out_tree(), out_parts))
-    c = xb.make_computation_builder('xla_computation_{}'.format(fun_name))
+    c = xb.make_computation_builder("xla_computation_{}".format(fun_name))
     xla_consts = map(partial(xb.constant, c), consts)
-    xla_args = xla._xla_callable_args(c, avals, tuple_args,
+    # As we cannot modify the closure-captured `tuple_args`, we work on a copy.
+    local_tuple_args = tuple_args
+    if local_tuple_args is None:
+      # pass long arg lists as tuple for TPU
+      local_tuple_args = len(avals) > 100
+    xla_args = xla._xla_callable_args(c, avals, local_tuple_args,
                                       partitions=in_parts_flat)
-    outs = xla.jaxpr_subcomp(
+    out_nodes = xla.jaxpr_subcomp(
         c, jaxpr, backend, axis_env_, xla_consts,
-        extend_name_stack(wrap_name(fun_name, 'xla_computation')), *xla_args)
-    build_out_tuple = partial(xc.ops.Tuple, c, outs)
+        extend_name_stack(wrap_name(fun_name, "xla_computation")), *xla_args)
+    build_out_tuple = partial(xc.ops.Tuple, c, out_nodes)
     if out_parts is not None:
       out_tuple = xb.with_sharding(c, out_parts_flat, build_out_tuple)
     else:
       out_tuple = build_out_tuple()
+
+    backend_ = xb.get_backend(backend)
+    if backend_.platform in ("gpu", "tpu"):
+      donated_invars = xla.set_up_aliases(c, xla_args, out_tuple,
+                                          donated_invars, tuple_args)
+    if any(donated_invars):
+      # TODO(tomhennigan): At call time we should mark these buffers as deleted.
+      unused_donations = [
+          str(c.GetShape(a)) for a, d in zip(xla_args, donated_invars) if d
+      ]
+      warn("Some donated buffers were not usable: {}".format(
+          ", ".join(unused_donations)))
     built = c.build(out_tuple)
     if return_shape:
```

Comment:
Could we remove the many choices of possible output format from the `_xla_computation` function? Suppose we always returned `_XlaReturn`, and the consumer (either `xla_computation` or `cpp_jit`) was responsible for unpacking the pieces they need.